### PR TITLE
Reinit layers of pretrained transformer in cached_transformers.get()

### LINKED ISSUE #5491: Re-initialize some layers of a PretrainedTransformerEmbedder
**Is your feature request related to a problem? Please describe.**

The paper [Revisiting Few-sample BERT Fine-tuning](https://arxiv.org/abs/2006.05987v3) (published at ICLR 2021) demonstrated that re-initializing the last few layers of a pretrained transformer before fine-tuning can reduce the variance between re-runs, speed up convergence and improve final task performance, nicely summarized in their figures:

image
image

The intuition is that some of the final layers may be over-specified to the pretraining objective(s) and therefore the pretrained weights can provide a bad initialization for downstream tasks. 

_It would be nice if re-initializing the the weights of certain layers in a pretrained transformer model was easy to do with AllenNLP._

**Describe the solution you'd like**

Ideally, you could easily specify which layers to re-initialize in a [`PretrainedTransformerEmbedder`](https://docs.allennlp.org/main/api/modules/token_embedders/pretrained_transformer_embedder/), something like:

```python
from allennlp.modules.token_embedders import PretrainedTransformerEmbedder

# Re-initialize the last 2 layers
embedder = PretrainedTransformerEmbedder(model_name="bert-base-uncased", reinit_layers=2)
# AND/OR, provide your own layer indices
embedder = PretrainedTransformerEmbedder(model_name="bert-base-uncased", reinit_layers=[10, 11])
```

The `__init__` of `PretrainedTransformerEmbedder` would take care of correctly re-initializing the specified layers for the given `model_name`.

**Describe alternatives you've considered**

You could achieve this right now with the [AllenNLP initializers](https://github.com/allenai/allennlp/blob/main/allennlp/nn/initializers.py), but this would require:

1. Writing regex to target each layer, which gets messy if you want to initialize some weights differently than others (like the weights/biases of LayerNorm vs FeedFoward).
2. Knowing how the model was initialized in the first place. E.g. BERT inits parameters using a truncated normal distribution with `mean=0` and `std=0.02`.  Ideally, the user wouldn't have to know/specify this.

**Additional context**

I've drafted a solution that works (but requires more testing). Essentially, we add a new parameter to `PretrainedTransformerEmbedder`, `reinit_layers`, which can be an integer or list of integers. In `__init__`, we re-initialize as follows:

```python
self._reinit_layers = reinit_layers
if self._reinit_layers and load_weights:
    num_layers = len(self.transformer_model.encoder.layer)
    if isinstance(reinit_layers, int):
        self._reinit_layers = list(range(num_layers - self._reinit_layers, num_layers))
    if any(layer_idx > num_layers for layer_idx in self._reinit_layers):
        raise ValueError(
            f"A layer index in reinit_layers ({self._reinit_layers}) is larger than the"
            f" maximum layer index {num_layers - 1}."
        )
    for layer_idx in self._reinit_layers:
        self.transformer_model.encoder.layer[layer_idx].apply(
            self.transformer_model._init_weights
        )
```

- Has no effect if `load_weights == False`, as the weights are already being randomly initialized.
- Takes advantage of the [`_init_weights`](https://github.com/huggingface/transformers/blob/d5d2744aa799b94488960a261d1b7376d791a621/src/transformers/modeling_bert.py#L592-L602) function from HF Transformers, which knows how to initialize the parameters of a layer correctly for a given pretrained model.
- A user can supply a list of layer indices, or a single integer, which we take to mean the last n layers.
- An error is raised if a layer index is greater than the maximum number of layers of the model.
- The default value of `reinit_layers` is None, so this should be backward compatible with existing configs.

I sanity-checked it by testing that the weights of the specified layers are indeed re-initialized. I also trained a model with re-initialized layers on my own task and got a non-negligible performance boost.

_If the AllenNLP maintainers think this would be a good addition I would be happy to open a PR!_

### PR DESCRIPTION
**Pull request by JohnGiorgi**
Fixes #5491.

Changes proposed in this pull request:

- Add a new parameter to `cached_transformers.get()`, `reinit_layers`.
- If `reinit_layers` is an integer, the parameters of the last `reinit_layers` of the transformer model are re-initialized.
- If `reinit_layers` is a list, the parameters of the layers indexed by `reinit_layers` of the transformer model are re-initialized.
- Added a test to make sure this feature works as expected.

## Before submitting

- [X] I've read and followed all steps in the [Making a pull request](https://github.com/allenai/allennlp/blob/main/CONTRIBUTING.md#making-a-pull-request)
    section of the `CONTRIBUTING` docs.
- [X] I've updated or added any relevant docstrings following the syntax described in the
    [Writing docstrings](https://github.com/allenai/allennlp/blob/main/CONTRIBUTING.md#writing-docstrings) section of the `CONTRIBUTING` docs.
- [X] If this PR fixes a bug, I've added a test that will fail without my fix.
- [X] If this PR adds a new feature, I've added tests that sufficiently cover my new functionality.

## After submitting

- [x] All GitHub Actions jobs for my pull request have passed.
- [x] **`codecov/patch`** reports high test coverage (at least 90%).
    You can find this under the "Actions" tab of the pull request once the other checks have finished.

### CODE REVIEW

**Comment by JohnGiorgi** [comment]
@epwalsh mind taking a look?

File: allennlp/modules/token_embedders/pretrained_transformer_embedder.py [hunk 1]
```diff
@@ -119,6 +126,22 @@ def __init__(
             self._scalar_mix = ScalarMix(self.config.num_hidden_layers)
             self.config.output_hidden_states = True
 
+        # Optionally, re-initialize the parameters of certain layers.
+        self._reinit_layers = cast(List[int], reinit_layers)
+        if self._reinit_layers and load_weights:
+            num_layers = len(self.transformer_model.encoder.layer)
```
Comment (epwalsh) [inline allennlp/modules/token_embedders/pretrained_transformer_embedder.py:132]: This should work fine for BERT, but won't for other models like XLM that name their modules differently. In general I think it'll be brittle to rely on guessing the internal module structure of transformer models.

But I can think of alternative which - although it's a little less friendly from a user perspective - still provides this useful functionality:
- Change the type of `reinit_layers` to `Optional[Union[int, List[int], List[str]]]`.
- When `reinit_layers` is `List[str]`, we interpret it as a list of regular expressions and we reinitialize all modules that match any of the regular expressions in the list.
- Otherwise when `reinit_layers` is `int` or `List[int]` we try the current approach, but if the given `transformer_model` does not have this module structure we throw a `ConfigurationError` and suggest that the user use the `List[str]` form of this parameter instead.

**Review by epwalsh — CHANGES_REQUESTED** [review]

File: allennlp/modules/token_embedders/pretrained_transformer_embedder.py (hunk 1 above)
Comment (JohnGiorgi) [reply allennlp/modules/token_embedders/pretrained_transformer_embedder.py]: Okay, I took a crack at this.

- I renamed `reinit_layers` to `reinit_modules` just because re-initializing modules (instead of whole layers) is possible when it is a list of regexes.
- Updated the type of the argument to `Optional[Union[int, List[int], List[str]]]`
- A configuration error is thrown if the model's layers cannot be easily indexed. In practice, most models I tried from hugging face could (BERTs, RoBERTa's) but you were right that XLM based models cant.
- Added tests for when `reinit_modules` is a list of strings.

Unfortunately, I cant seem to get the weights to re-initialize. I figured that after finding a match between the regex and some `module` name I could just call `module.apply(self.transformer_model._init_weights)` like so

```python
for regex in self._reinit_modules:
    for name, module in self.transformer_model.named_modules():
        if re.search(regex, name):
            module.apply(self.transformer_model._init_weights)
```

but this doesn't actually re-init the `module`s weights? Strange. I will have to dig into the HF repo and docs to figure out why.

**Comment by JohnGiorgi** [comment]
@epwalsh Okay, I believe everything works and this new functionality is covered by a bunch of tests. I think the only outstanding questions are:

- This code has gotten quite long and complicated. It might be better to place it in its own method rather than `__init__`.
- In the unit test, I used `"xlm-mlm-enfr-1024"` as the XLM model to test, but it is quite big. Might be worth it to look for a smaller XLM model.
- I am realizing now that it might have made more sense to add this logic to `cached_transformers`, so that it can be used outside of just `PretrainedTransformerEmbedder`. If you agree I can move it over.

**Comment by epwalsh** [comment]
> I am realizing now that it might have made more sense to add this logic to cached_transformers, so that it can be used outside of just PretrainedTransformerEmbedder. If you agree I can move it over.

That makes sense to me!

> In the unit test, I used "xlm-mlm-enfr-1024" as the XLM model to test, but it is quite big. Might be worth it to look for a smaller XLM model.

We could test with a GPT2 model instead since those don't have an `encoder` module either. And there is https://huggingface.co/sshleifer/tiny-gpt2.

**Comment by JohnGiorgi** [comment]
> > I am realizing now that it might have made more sense to add this logic to cached_transformers, so that it can be used outside of just PretrainedTransformerEmbedder. If you agree I can move it over.
> 
> That makes sense to me!

Decided against this as I realized it's actually a little complicated. The caching doesn't work because `reinit_modules` can be a list and that isn't hashable. Leaving this functionality in `PretrainedTransformerEmbedder`

> > In the unit test, I used "xlm-mlm-enfr-1024" as the XLM model to test, but it is quite big. Might be worth it to look for a smaller XLM model.
> 
> We could test with a GPT2 model instead since those don't have an `encoder` module either. And there is https://huggingface.co/sshleifer/tiny-gpt2.

Perfect! I used tiny-gpt2 and it made the tests run quite a bit faster

**Comment by epwalsh** [comment]
> Decided against this as I realized it's actually a little complicated. The caching doesn't work because reinit_modules can be a list and that isn't hashable.

We could convert it to a tuple instead, which will be hashable

File: allennlp/common/cached_transformers.py [hunk 2]
```diff
@@ -110,6 +138,40 @@ def strip_prefix(s):
...
+                        f" {model_name} using layer indices. Please provide a list of"
+                        " strings corresponding to the names of the layers to re-initialize."
+                    )
+            elif all(isinstance(x, str) for x in reinit_modules):
+                for regex in reinit_modules:
+                    for name, module in transformer.named_modules():
+                        if re.search(regex, name):
+                            module.apply(transformer._init_weights)
+            else:
+                raise ValueError("reinit_modules must be a list of strings or a list of integers.")
```
Comment (epwalsh) [inline allennlp/common/cached_transformers.py:174]:
```suggestion
                raise ValueError("reinit_modules must be either an integer, a list of strings, or a list of integers.")
```

File: allennlp/common/cached_transformers.py [hunk 3]
```diff
@@ -43,13 +46,26 @@ def get(
...
     override_weights_strip_prefix : `str`, optional (default = `None`)
         If set, strip the given prefix from the state dict when loading it.
+    reinit_modules: `Optional[Union[int, Tuple[int, ...], Tuple[str, ...]]]`, optional (default = `None`)
+        If this is an integer, the last `reinit_modules` layers of the transformer will be
+        re-initialized. If this is a tuple of integers, the layers indexed by `reinit_modules` will
+        be re-initialized. If this is a tuple of strings, they will be treated as regexes and any
+        module with a name matching the regex will be re-initialized. Re-initializing the last few
+        layers of a pretrained transformer can reduce the instability of fine-tuning on small
+        datasets and may improve performance (https://arxiv.org/abs/2006.05987v3). Has no effect
+        if `load_weights` is `False` or `override_weights_file` is not None.
```
Comment (epwalsh) [inline allennlp/common/cached_transformers.py:56]: I think it would be helpful to have a note here about how we can't guarantee using the `int` or `list[int]` form will work. And if that fails, the user should use the `list[str]` form.

**Review by epwalsh — COMMENTED** [review]
This looks really great! I just have a couple minor comments

File: allennlp/common/cached_transformers.py (hunk 3 above)
Comment (JohnGiorgi) [reply allennlp/common/cached_transformers.py]: Good call! Updated this docstring in `PretrainedTransformerEmbedder` and `cached_transformers.get()` with a note warning that layer indices could fail

File: allennlp/common/cached_transformers.py [hunk 4]
```diff
@@ -43,13 +46,28 @@ def get(
         with `torch.save()`.
     override_weights_strip_prefix : `str`, optional (default = `None`)
         If set, strip the given prefix from the state dict when loading it.
+    reinit_modules: `Optional[Union[int, Tuple[int, ...], Tuple[str, ...]]]`, optional (default = `None`)
+        If this is an integer, the last `reinit_modules` layers of the transformer will be
+        re-initialized. If this is a tuple of integers, the layers indexed by `reinit_modules` will
+        be re-initialized. Note, because the module structure of the transformer `model_name` can
+        differ, we cannot guarantee that providing an integer or tuple of integers will work. If
+        this fails, you can instead provide a tuple of strings, which will be treated as regexes and
+        any module with a name matching the regex will be re-initialized. Re-initializing the last
```
Comment (JohnGiorgi) [inline allennlp/common/cached_transformers.py:55]: @epwalsh Updated part of docstring here

**Review by epwalsh — APPROVED** [review]
LGTM! I think https://github.com/allenai/allennlp/pull/5529 should fix the failures in CI, so once that goes through and all tests pass I'll merge. Thanks @JohnGiorgi!

**Comment by JohnGiorgi** [comment]
> LGTM! I think #5529 should fix the failures in CI, so once that goes through and all tests pass I'll merge. Thanks @JohnGiorgi!

Awesome :) Thanks for your help as always!!

**Comment by dirkgr** [comment]
@JohnGiorgi, @epwalsh, I don't think this makes a ton of sense as implemented. You want to cache the weights because you want to re-use them in many places. But `_init_weights()` randomizes the weights. The way it is now, you will get the same randomized weights every time. Even if that's what you want, randomizing the weights only makes sense if you train (i.e., alter) them later. When you do that, you need to make a copy of the weights, so you don't change the cached version (and thus affect other users of the cache).

We can still have this be part of `cached_transformers.get()`, but we should not cache the version with re-initialized weights. `reinit_layers` should not be part of `TransformerSpec`. The re-initialization should happen every time `get()` is called.

**Comment by JohnGiorgi** [comment]
@dirkgr I see, good catch. Opened a PR with those fixes here: https://github.com/allenai/allennlp/pull/5543

**Comment by dirkgr** [comment]
Thanks!
